src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,335 @@
+            }
+
+            if (blocks == null)
+            {
+                throw new ArgumentNullException(nameof(blocks));
+            }
+
+            if (callbackExecuter == null)
+            {
+                throw new ArgumentNullException(nameof(callbackExecuter));
+            }
+
+            this.callbackRepository = callbackRepository;
+            this.watchRepository = watchRepository;
+            this.callbackExecuter = callbackExecuter;
+
+            this.watcher = new Zcoin.Watching.TransactionConfirmationWatcher<ConfirmContext>
+            (
+                this,
+                blocks
+            );
+
+            timerLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+            timers = new Dictionary<uint256, Dictionary<Guid, Tuple<Threading.Timer, ConfirmContext>>>();
+            watches = new ConcurrentDictionary<Guid, TransactionWatch<ConfirmContext>>();
+        }
+
+        public async Task<ConfirmContext> AddTransactionAsync(
+            uint256 transaction,
+            int confirmation,
+            TimeSpan timeout,
+            IPAddress registeringIp,
+            Uri callbackUrl,
+            TransactionConfirmationCallbackResult successData,
+            TransactionConfirmationCallbackResult timeoutData,
+            CancellationToken cancellationToken)
+        {
+            var callback = await this.callbackRepository.AddAsync
+            (
+                registeringIp,
+                callbackUrl,
+                cancellationToken
+            );
+
+            var watch = await this.watchRepository.AddAsync
+            (
+                transaction,
+                confirmation,
+                timeout,
+                successData,
+                timeoutData,
+                callback,
+                cancellationToken
+            );
+
+            TimerSetup(watch);
+
+            return watch;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var allWatches = await this.watchRepository.ListAsync(cancellationToken);
+
+            foreach (var watch in allWatches.Where(w => !w.Callback.Completed))
+            {
+                TimerSetup(watch);
+            }
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            timerLock.EnterWriteLock();
+
+            try
+            {
+                foreach (var timerSet in timers)
+                {
+                    foreach (var timer in timerSet.Value)
+                    {
+                        await timer.Value.Item1.StopAsync(cancellationToken);
+                    }
+                }
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+        }
+
+        Task IBlockListener.BlockAddedAsync(Block block, int height, CancellationToken cancellationToken)
```

Comment:
The explicit implementation should come after other methods due to it not directly callable within the class so all callable will grouped together.